File: balanced_forest/balanced_forest.py
#!/bin/python3

import math
import os
import random
import re
import sys

from collections import deque


def get_graph(edges):
    graph = {}
    for edge in edges:
        node_a, node_b = edge[0] - 1, edge[1] - 1
        graph[node_a] = graph.get(node_a, set())
        graph[node_b] = graph.get(node_b, set())
        graph[node_a].add(node_b)
        graph[node_b].add(node_a)
    return graph


def get_tree(graph):
    leaves = [node for node in graph if len(graph[node]) == 1]
    q = deque(leaves)
    tree = {}
    nodes = set(graph.keys())
    while len(graph) > 2:
        for _ in range(len(q)):
            leaf = q.popleft()
            for neighbor in graph[leaf]:
                graph[neighbor].remove(leaf)
                tree[leaf] = neighbor
                if len(graph[neighbor]) == 1:
                    q.append(neighbor)
            del graph[leaf]
    left_nodes = list(graph.keys())
    if len(left_nodes) == 2:
        root = left_nodes[1]
        tree[left_nodes[0]] = root
    else:
        root = left_nodes[0]
    return tree, nodes, root

def get_sums_and_children(tree, nodes, root, c):
    sums = {}
    children = {}
    for node in nodes:
        aux = node
        while True:  # aux != root
            try:
                # For sums
                sums[aux] = sums.get(aux, 0) + c[node]
                # For children
                children[aux] = children.get(aux, set())
                if node != aux:
                    children[aux].add(node)
                aux = tree[aux]
            except KeyError:  # aux == root -> root not in tree
                break
    return sums, children

def balancedForest(c, edges):
    tree, nodes, root = get_tree(get_graph(edges))
    sums, children = get_sums_and_children(tree, nodes, root, c)
    val = float("inf")
    checked = {}
    for n1 in children[root]:
        valid_sum = {n1: sums[n1], root: sums[root] - sums[n1]}
        node_lo, node_hi = sorted(valid_sum, key=lambda k: valid_sum[k])
        if valid_sum[node_hi] == valid_sum[node_lo]:
            val = min(val, valid_sum[node_lo])
            continue
        if valid_sum[node_hi] <= 2 * valid_sum[node_lo]:
            goal_sum = valid_sum[node_lo]
            new_node_val = 2 * valid_sum[node_lo] - valid_sum[node_hi]
            if node_hi == n1:
                curr_children = children[n1]
            else:
                curr_children = children[root] - children[n1].union([n1])
        elif valid_sum[node_hi] > 2 * valid_sum[node_lo]:
            if valid_sum[node_hi] % 2 == 0:
                goal_sum = valid_sum[node_hi] // 2
                new_node_val = goal_sum - valid_sum[node_lo]
                if node_hi == n1:
                    curr_children = children[n1]
                else:
                    curr_children = children[root] - children[n1].union([n1])
            else:
                continue

        for n2 in curr_children:
            if (n2, n1) not in checked:
                checked[n1, n2] = True
                sum_bottom = sums[n2]
                sum_top = valid_sum[node_hi] - sum_bottom
                if sum_top == goal_sum or sum_bottom == goal_sum:
                    val = min(val, new_node_val)
                    break
    return val if val != float("inf") else -1


if __name__ == '__main__':
    with open("input05.txt", "r") as file:
        q = int(file.readline())
        for q_itr in range(q):
            n = int(file.readline())

            c = list(map(int, file.readline().rstrip().split()))

            edges = []

            for _ in range(n - 1):
                edges.append(list(map(int, file.readline().rstrip().split())))

            result = balancedForest(c, edges)

            print(result)

File: balanced_forest/balanced_forest.cpp
#include <bits/stdc++.h>
using namespace std;

unordered_map<int, unordered_set<int>> &
get_graph(vector<vector<int>> & edges) {
    unordered_map<int, unordered_set<int>> graph;
    for (vector<int> edge : edges) {
        int node_a = edge[0] - 1, node_b = edge[1] - 1; 
        graph[node_a] = graph.find(node_a) != graph.end() ? graph[node_a] : unordered_set<int>();
        graph[node_b] = graph.find(node_b) != graph.end() ? graph[node_b] : unordered_set<int>();
        graph[node_a].insert(node_b);
        graph[node_b].insert(node_a);
    }
    return graph;
}

tuple<unordered_map<int, int>, unordered_set<int>, int>
get_tree(unordered_map<int, unordered_set<int>> & graph) {
    queue<int> leaves_q;
    unordered_set<int> nodes;
    for (auto node_neighbors : graph) {
        if (node_neighbors.second.size() == 1) {
            leaves_q.push(node_neighbors.first);
        }
        nodes.insert(node_neighbors.first);
    }
    unordered_map<int, int> tree;
    while (graph.size() > 2) {
        int q_len = leaves_q.size();
        for (int i = 0; i < q_len; i++) {
            int leaf = leaves_q.front();
            leaves_q.pop();
            for (int neighbor : graph[leaf]) {
                graph[neighbor].erase(leaf);
                tree[leaf] = neighbor;
                if (graph[neighbor].size() == 1){
                    leaves_q.push(neighbor);
                }
            }
            graph.erase(leaf);
        }
    }
    vector<int> left_nodes;
    for (auto node_neighbors: graph) left_nodes.push_back(node_neighbors.first);
    int root;
    if (left_nodes.size() == 2) {
        root = left_nodes[1];
        tree[left_nodes[0]] = root;
    }
    else {
        root = left_nodes[0];
    }
    return make_tuple(tree, nodes, root);;
}

tuple<unordered_map<int, long>, unordered_map<int, unordered_set<int>>>
get_sums_and_children(unordered_map<int, int> tree, unordered_set<int> nodes, int root, vector<int> c) {
    unordered_map<int, long> sums;
    unordered_map<int, unordered_set<int>> children;
    for (int node : nodes) {
        int aux = node;
        while (true) { // aux != root
            try {
                // For sums
                sums[aux] = ((sums.find(aux) != sums.end()) ? sums[aux] : 0) + c[node];
                // For children
                children[aux] = ((children.find(aux) != children.end()) ? children[aux] : unordered_set<int>());
                if (node != aux) {
                    children[aux].insert(node);
                }
                aux = tree[aux];
            } catch (out_of_range oor) { // aux == root -> root not in tree
                break;
            }
        }  
    }
    return make_tuple(sums, children);
}

long balancedForest(vector<int> c, vector<vector<int>> edges) {
    auto tree_components = get_tree(get_graph(edges));
    unordered_map<int, int> tree = get<0>(tree_components);
    unordered_set<int> nodes = get<1>(tree_components);
    int root = get<2>(tree_components);
    auto sums_and_children = get_sums_and_children(tree, nodes, root, c);
    unordered_map<int, long> sums = get<0>(sums_and_children);
    unordered_map<int, unordered_set<int>> children = get<1>(sums_and_children);
    long val = LONG_MAX;
    vector<vector<int>> checked(c.size(), vector<int>(c.size(), 0));
    for (auto n1: children[root]) {
        map<int, long> valid_sum;
        valid_sum[n1] = sums[n1];
        valid_sum[root] = sums[root] - sums[n1];
        int node_lo = valid_sum[n1] < valid_sum[root] ? n1 : root;
        int node_hi = valid_sum[n1] < valid_sum[root] ? root : n1;
        if (valid_sum[node_hi] == valid_sum[node_lo]){
            val = min(val, valid_sum[node_lo]);
            continue;
        }
        long goal_sum, new_node_val;
        unordered_set<int> curr_children;
        if (valid_sum[node_hi] < 2 * valid_sum[node_lo]) {
            goal_sum = valid_sum[node_lo];
            new_node_val = 2 * valid_sum[node_lo] - valid_sum[node_hi];
            if (node_hi == n1) {
                curr_children = children[n1];
            }
            else {
                curr_children.clear();
                copy_if(
                    children[root].begin(),
                    children[root].end(), 
                    inserter(curr_children, curr_children.end()), 
                    [&children, n1] (int node) { return (children[n1].find(node) == children[n1].end() || node != n1); }
                );
            }
        }
        else if (valid_sum[node_hi] > 2 * valid_sum[node_lo]){
            if (valid_sum[node_hi] % 2 == 0){
                goal_sum = valid_sum[node_hi] / 2;
                new_node_val = goal_sum - valid_sum[node_lo];
                if (node_hi == n1) {
                    curr_children = children[n1];
                }
                else{
                    curr_children.clear();
                    copy_if(
                        children[root].begin(),
                        children[root].end(), 
                        inserter(curr_children, curr_children.end()), 
                        [&children, n1] (int node) { return (children[n1].find(node) == children[n1].end() || node != n1); }
                    );
                }
            }
            else {
                continue;
            }
        }
        else { // valid_sum[node_hi] == 2 * valid_sum[node_lo]
            goal_sum = valid_sum[node_hi] / 2;
            new_node_val = 0;  // goal_sum - valid_sum[node_lo]
            if (node_hi == n1) {
                curr_children = children[n1];
            }
            else {
                curr_children.clear();
                copy_if(
                    children[root].begin(),
                    children[root].end(), 
                    inserter(curr_children, curr_children.end()), 
                    [&children, n1] (int node) { return (children[n1].find(node) == children[n1].end() || node != n1); }
                );
            }
        }  
        for (auto n2 :curr_children){
            if (!checked[n2][n1]) {
                checked[n1][n2] = 1;
                long sum_bottom = sums[n2];
                long sum_top = valid_sum[node_hi] - sum_bottom;
                if (sum_top == goal_sum or sum_bottom == goal_sum) {
                    val = min(val, new_node_val);
                    break;
                }
            }
        }
    }
    return val != LONG_MAX ? val : -1;
}
